docs/TICCL-lexclean.1
.TH TICCL-lexclean 1 "2020 mar 09"

.SH NAME
TICCL-lexclean - a lexicon cleaning tool. Note this tool does not form part of the TICCL pipeline proper!

.SH SYNOPSIS

TICCL-lexclean [options] FILES

.SH DESCRIPTION
.B TICC-lexclean
processes a tab separated frequency file (for instance a *.tsv file produced by
.B FoLiA-stats or
.B TICCL-stats or
) and filters out non-alphabet words.

.SH OPTIONS
.B -a
file
.RS
name of the alphabet file. This are the characters to accept.
This file must be in the format provided by
.B TICCL-lexstat.
.RE

.B -x
file
.RS
the reverse of -a. This file contains all characters to reject.
.RE

.B -t
.RS
assume the input contains POS tags too.
.RE

.B -p
.RS
output percentages too.
.RE

.B -V or
--version
.RS
Show VERSION
.RE

.B -h
or
.B --help
.RS
usage info
.RE

.SH BUGS
possibly

.SH AUTHORS
Ko van der Sloot lamasoftware@science.ru.nl

Martin Reynaert reynaert@uvt.nl

.SH SEE ALSO
.BR TICCL-lexstat (1)
.BR FoLiA-stats (1)


NEWS
ticcltools 0.6 2018-06-05
[Ko vander Sloot]
Intermediate release, with a lot of new code to handle N-grams
Also a lot of refactoring is done, for more clear and maintainable code.
This is work in progress still.

* TICCL-unk:
  - more extensive acronym detection
  - fixed artifreq problems in 'clean' punctuated words
  - added filters for 'unwanted' characters
  - added a ligature filter to convert evil ligatures
  - normalize all hyphens to a 'normal' one (-)
  - use a better definition of punctuation (unicode character class is not
    good enough to decide)

* TICCL-lexstat:
  - the 'separator' symbol should get freq=0, so it isn't counted
  - the clip value is added to the output filename

* TICCL-indexer:
  - indexer and indexerNT now produce the same output, using different
    strategies when a --foci files is used.

* TICCL-LDcalc:
  major overhaul for n-grams
  - added a ngram point column to the output (so NOT backward compatible!)
  - produce a '.short' list for short word corrections
  - produce a '.ambi' file with a list of n-grams related to short words
  - prune a lot of ngrams from the output

* TICCL-rank:
 - output is sorted now
 - honor the ngram-points from the new LDcalc. (so NOT backward compatible!)

* TICCL-chain: new module to chain ranked files

* TICCL-lexclean:
  -added a -x option for 'inverse' alphabet

* TICCL-anahash:
  - added a --list option to produce a list of words and anagram values

[Maarten van Gompel]
* added metadata file: codemeta.json

ticcltools 0.5 2018-02-19
[Ko van der Sloot]
* updated configuration. also for Mac OSX
* use of more ticcutils stuff: diacritics filter
* added a TICCL-mergelex program
* the OMP_THREAD_LIMIT environment variable was ignored sometimes
* TICCL-unk:
   - fixed a problem in artifreq handling
   - changed acronym detection (work in  progress)
   - added -o option
  TICCL-lexstat:
   - added TTR output
   - added -o option
  TICCL-indexer
  - now also handles --foci file. with some speed-up
  - added a -t option
  TICCL-LDcalc:
  - be less picky on a few wrong lines in the data
* added some tests
* when libroaring is installed we built roaring versions of some modules (experimental)
* updated man pages

ticcltools 0.4 2017-04-04
[Ko van der Sloot]
* first official release.
  - added functions to test on Word2Vec datafiles
  - refactoring and modernizing stuff all around

ticcltools 0.3 2016-01-21
* upload ready
ticcltools 0.2 2016-01-14
[Ko van der Sloot]
* repository moved to GitHub
* Travis support
* first 'distributable' version
* added TICCL-stats program
* added W2V-near and W2V-dist programs bases on new libticcl lib


[0.1] Ko vd Sloot
 started autoconfiscating


src/TICCL-lexclean.cxx
/*
  Copyright (c) 2006 - 2020
  CLST  - Radboud University
  ILK   - Tilburg University

  This file is part of ticcltools

  ticcltools is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 3 of the License, or
  (at your option) any later version.

  ticcltools is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, see <http://www.gnu.org/licenses/>.

  For questions and suggestions, see:
      https://github.com/LanguageMachines/ticcltools/issues
  or send mail to:
      lamasoftware (at ) science.ru.nl

*/

#include <cstdlib>
#include <string>
#include <map>
#include <vector>
#include <iomanip>
#include <iostream>
#include <fstream>

#include "ticcutils/CommandLine.h"
#include "ticcutils/StringOps.h"
#include "ticcutils/FileUtils.h"
#include "ticcutils/Unicode.h"

#include "config.h"

using namespace	std;
using namespace	icu;

void create_wf_list( const map<string, unsigned int>& wc,
		     const string& filename, unsigned int totalIn,
		     bool doperc ){
  unsigned int total = totalIn;
  ofstream os( filename );
  if ( !os ){
    cerr << "failed to create outputfile '" << filename << "'" << endl;
    exit(EXIT_FAILURE);
  }
  map<unsigned int, set<string> > wf;
  for ( const auto& cit : wc ){
    wf[cit.second].insert( cit.first );
  }
  unsigned int sum=0;
  map<unsigned int, set<string> >::const_reverse_iterator wit = wf.rbegin();
  while ( wit != wf.rend() ){
    if ( wit->first == 0 ){
      for ( const auto& s : wit->second ){
	os << s << endl;
	++total;
      }
    }
    else {
      for ( const auto& s : wit->second ){
	sum += wit->first;
	os << s << "\t" << wit->first;
	if ( doperc ){
	  os << "\t" << sum << "\t" << std::setprecision(8) << 100 * double(sum)/total;
	}
	os << endl;
      }
    }
    ++wit;
  }
  cout << "created cleaned list '" << filename << "'" << endl;
  cout << "with " << total << " words." << endl;
}

void dump_quarantine( const string& filename,
		      const map<string, unsigned int>& qw ){
  ofstream os( filename );
  if ( !os ){
    cerr << "failed to create outputfile '" << filename << "'" << endl;
    exit(EXIT_FAILURE);
  }
  for ( const auto& it : qw ){
    os << it.first;
    if ( it.second > 0 ){
      os << "\t" << it.second;
    }
    os << endl;
  }
  cout << "created quarantine list '" << filename << "'" << endl;
  cout << "with " << qw.size() << " items. " << endl;
}

bool isClean( const string& s, const set<UChar>& alp, bool reverse ){
  UnicodeString us = TiCC::UnicodeFromUTF8( s );
  //  cerr << "check " << us << endl;
  for ( int i=0; i < us.length(); ++i ){
    //    cerr << "check " << us[i] << endl;
    if ( alp.find( us[i] ) == alp.end() ){
      if ( reverse ){
	continue;
      }
      //      cerr << "rejected" << endl;
      return false;
    }
    if ( reverse ){
      //      cerr << "rejected" << endl;
      return false;
    }
  }
  //  cerr << "OK" << endl;
  return true;
}

bool fillAlpha( const string& file, set<UChar>& alphabet ){
  string line;
  ifstream is( file );
  while ( getline( is, line ) ){
    if ( line.size() == 0 || line[0] == '#' ){
      continue;
    }
    vector<string> v = TiCC::split( line );
    UnicodeString us = TiCC::UnicodeFromUTF8( v[0] );
    us.toLower();
    alphabet.insert( us[0] );
    us.toUpper();
    alphabet.insert( us[0] );
    // for now, we don't use the other fields
  }
  return true;
}


void usage(){
  cerr << "Usage: [options] file/dir" << endl;
  cerr << "\t TICCL-lexclean will clean 1, 2 or 4 columned (lexicon) files" << endl;
  cerr << "\t The output will be a 1, 2 or 4 columned tab separated file, extension: .cleaned " << endl;
  cerr << "\t\t (4 columns when -p is specified)" << endl;
  cerr << "\t 'dirty' words are written to a .dirty file." << endl;
  cerr << "\t-a\t alphabet file. With characters to accept" << endl;
  cerr << "\t-x\t unwanted alphabet file. With characters to reject" << endl;
  cerr << "\t-p\t output percentages too. " << endl;
  cerr << "\t-t\t assume a POS tagged input" << endl;
  cerr << "\t-h or --help\t this message" << endl;
  cerr << "\t-V or --version\t show version " << endl;
}

int main( int argc, char *argv[] ){
  TiCC::CL_Options opts( "hVpa:x:t", "help,version" );
  try {
    opts.init(argc,argv);
  }
  catch( TiCC::OptionError& e ){
    cerr << e.what() << endl;
    usage();
    exit( EXIT_FAILURE );
  }

  bool dopercentage = false;
  bool postagged = false;
  string value;
  string alpha;
  string no_alpha;
  bool reverse = false;
  if ( opts.extract('V', value ) || opts.extract("version",value) ){
    cerr << PACKAGE_STRING << endl;
    exit(EXIT_SUCCESS);
  }
  if ( opts.extract('h', value ) || opts.extract("help", value ) ){
    usage();
    exit(EXIT_SUCCESS);
  }
  if ( opts.extract('a', value ) ){
    alpha = value;
  }
  if ( opts.extract('x', value ) ){
    no_alpha = value;
    reverse = true;
  }
  if ( !alpha.empty() && !no_alpha.empty() ){
    cerr << "may not combine -a and -x" << endl;
    exit( EXIT_FAILURE );
  }
  if ( opts.extract('p', value ) ){
    dopercentage = true;
  }
  if ( opts.extract('t', value ) ){
    postagged = true;
  }
  if ( !opts.empty() ){
    usage();
    exit(EXIT_FAILURE);
  }

  set<UChar> alphabet;
  if ( !alpha.empty() ){
    fillAlpha( alpha, alphabet );
    cerr << "read alphabet file with " << alphabet.size()
	 << " characters" << endl;
  }
  if ( !no_alpha.empty() ){
    fillAlpha( no_alpha, alphabet );
    cerr << "read EXCLUDE alphabet file with " << alphabet.size()
	 << " characters" << endl;
  }

  vector<string> fileNames = opts.getMassOpts();

  size_t toDo = fileNames.size();
  if ( toDo == 0 ){
    cerr << "no matching files found" << endl;
    usage();
    exit(EXIT_SUCCESS);
  }
  map<string,unsigned int> wc;
  map<string,unsigned int> qw;
  for ( const auto& docName : fileNames ){
    ifstream is( docName );
    string line;
    unsigned int word_total = 0;
    while ( getline( is, line ) ){
      vector<string> vec = TiCC::split_at( line, "\t" );
      size_t num = vec.size();
      if ( num == 1 ){
	if ( isClean( vec[0], alphabet, reverse ) ){
	  wc[vec[0]] = 0;
	}
	else {
	  qw[vec[0]] = 0;
	}
      }
      else if ( num == 2 || num == 4 ){
	string val = vec[0];
	vector<string> v2;
	if ( postagged ){
	  if ( TiCC::split( val, v2 ) > 1 ){
	    val = v2[0];
	  }
	  else {
	    cerr << "pos tagged files need a space separated value in the first column" << endl;
	    exit(EXIT_FAILURE);
	  }
	}
	unsigned int freq = TiCC::stringTo<unsigned int>( vec[1] );
	if ( isClean( val, alphabet, reverse ) ){
	  wc[vec[0]] = freq;
	  word_total += freq;
	}
	else {
	  qw[vec[0]] = freq;
	}
      }
      else {
	cerr << "unexpected line: '" << line << "' in " << docName << endl;
	continue;
      }
    }
    string outname = docName + ".cleaned";
    create_wf_list( wc, outname, word_total, dopercentage );
    outname = docName + ".dirty";
    dump_quarantine( outname, qw );
  }
}
